specutils/io/default_loaders/sdss.py

```diff
@@ -26,29 +27,74 @@
 
 def spec_identify(origin, *args, **kwargs):
     """
-    Check whether given file is FITS and has SDSS-III/IV spec type
+    Check whether given input is FITS and has SDSS-III/IV spec type
     BINTABLE in first extension. This is used for Astropy I/O Registry.
     """
-    with fits.open(args[0]) as hdulist:
-        # Test if fits has extension of type BinTable and check for
-        # spec-specific keys
-        return (hdulist[0].header['TELESCOP'] == 'SDSS 2.5-M' and
-                len(hdulist) > 1 and
-                isinstance(hdulist[1], fits.BinTableHDU) and
-                hdulist[1].header['TTYPE3'] == 'ivar')
+    if isinstance(args[2], fits.hdu.hdulist.HDUList):
+        hdulist = args[2]
+    elif isinstance(args[2], _io.BufferedReader):
+        hdulist = fits.open(args[2])
+    else:
+        hdulist = fits.open(args[0], **kwargs)
+
+    # Test if fits has extension of type BinTable and check for spec-specific keys
+    is_sdss = (hdulist[0].header['TELESCOP'] == 'SDSS 2.5-M' and
+               hdulist[0].header.get('FIBERID', 0) > 0 and
+               len(hdulist) > 1 and
+               isinstance(hdulist[1], fits.BinTableHDU) and
+               hdulist[1].header['TTYPE3'] == 'ivar')
+
+    if not isinstance(args[2], (fits.hdu.hdulist.HDUList, _io.BufferedReader)):
+        hdulist.close()
+    return is_sdss
 
 
 def spSpec_identify(origin, *args, **kwargs):
     """
-    Check whether given file is FITS with SDSS-I/II spSpec type data.
+    Check whether given input is FITS with SDSS-I/II spSpec tyamepe data.
     This is used for Astropy I/O Registry.
     """
-    with fits.open(args[0]) as hdulist:
-        # Test telescope keyword and check if primary HDU contains data
-        # consistent with spSpec format
-        return (hdulist[0].header['TELESCOP'] == 'SDSS 2.5-M' and
-                isinstance(hdulist[0].data, np.ndarray) and
-                hdulist[0].data.shape[0] == 5)
+    if isinstance(args[2], fits.hdu.hdulist.HDUList):
+        hdulist = args[2]
+    elif isinstance(args[2], _io.BufferedReader):
+        hdulist = fits.open(args[2])
+    else:
+        hdulist = fits.open(args[0])
```

Comment:
What makes the location of the hdulist in `args` non-deterministic?

Reply:
`asdf 2.6.0` added a big validator efficiency on read.  `jwst.datamodels` specifically does not want to use that, and there was a period of about 2 weeks were our last release was not compatible with the latest `asdf 2.6.0`.  I.e. serialized files would be written out incorrectly.

Reply:
It's not really non-deterministic, but possibly a bit confusing from the calling sequence in `astropy.io.registry.identify_format` – according to the docs and as it appears also from the code, it should be calling the identifier with
`_identifiers[(data_format, data_class)](origin, path, fileobj, *args, **kwargs)`
thus the filename will be passed in `args[0]`; for `open(filename, mode='rb')` I found it to pass `args[:3]` as
```
None
<_io.BufferedReader name='spec-1323-52797-0012.fits'>
<_io.BufferedReader name='spec-1323-52797-0012.fits'>
```
whereas for the hdulist this becomes
```
None
None
[<astropy.io.fits.hdu.image.PrimaryHDU object at 0x1102962e0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x1103552e0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x110355520>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x110355760>]
```
so apparently the FITS object is moved to `args[2]`, while a regular `fileobj` comes with copies in `args[1]` and `args[2]`. But perhaps it would be more consistent with the `registry` documentation to make the 3rd case
```
    elif isinstance(args[1], _io.BufferedReader):
        hdulist = fits.open(args[1])
```
(it works both ways).

Reply:
Strange. I'm not too worried about it, but it seems that the specutils documentation on custom loaders ought to be updated to reflect some of these nuances for new users.

Reply:
Yes, I think I would have had a hard time figuring any of this out without the earlier primer you gave me on using `identify`, and the examples in #637. https://docs.astropy.org/en/stable/io/registry.html#using-astropy-io-registry does not give a great lot of detail either, but I think summarising the main aspects relevant to `specutils` usage would make a good addition to the docs.
Also in principle most other loaders could probably also be enabled to auto-detect fileobj input.